File: 한민선/240604/1991. 트리 순회/트리 순회.java
import java.io.*;
import java.util.*;

public class Main {
    static Node head = new Node('A', null, null);

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine());
        for (int i = 0; i < t; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            char root = st.nextToken().charAt(0);
            char left = st.nextToken().charAt(0);
            char right = st.nextToken().charAt(0);
            // head 노드를 기준으로 입력받은 노드를 붙임
            addNode(head, root, left, right);
        }

        preOrder(head);
        System.out.println();
        inOrder(head);
        System.out.println();
        postOrder(head);
    }
    static class Node{
        char value;
        Node left;
        Node right;

        public Node(char value, Node left, Node right) {
            this.value = value;
            this.left = left;
            this.right = right;
        }
    }
    public static void addNode(Node target, char root, char left, char right){
        // root인 경우(가장 첫번째 노드) -> root의 왼쪽과 오른쪽 노드 값을 넣어줌
        if(target.value == root){
            target.left = (left == '.' ? null : new Node(left, null, null));
            target.right = (right == '.' ? null : new Node(right, null, null));
        }
        else {
            // 하위노드가 있는지 확인 후 추가
            if(target.left != null) addNode(target.left, root, left, right);
            if(target.right != null) addNode(target.right, root, left, right);
        }
    }

    // 전위 순회 VLR
    public static void preOrder(Node node){
        if(node == null) return;
        System.out.print(node.value);
        preOrder(node.left);
        preOrder(node.right);
    }
    // 중위 순회 LVR
    public static void inOrder(Node node){
        if (node == null) return;
        inOrder(node.left);
        System.out.print(node.value);
        inOrder(node.right);
    }
    // 후위 순회 LRV
    public static void postOrder(Node node){
        if (node == null) return;
        postOrder(node.left);
        postOrder(node.right);
        System.out.print(node.value);
    }
}

File: 이하영/240604/1991. 트리 순회/트리 순회.py
# 전위 : 루트 -> 왼쪽 -> 오른쪽
# 중위 : 왼쪽 -> 루트 -> 오른쪽
# 후위 : 왼쪽 -> 오른쪽 -> 루트

# 입력 -> 본인을 key으로, 자식 노드(왼쪽 오른쪽) value
import sys

N = int(sys.stdin.readline())

tree = dict()

for _ in range(N):
    root, left, right = sys.stdin.readline().split()
    tree[root] = [left, right]
    
def preorder(node): #전위
    print(node, end='') #루트부터 방문
    if tree[node][0] != ".":
        preorder(tree[node][0]) #왼쪽 탐색
    if tree[node][1] != ".":
        preorder(tree[node][1]) #오른쪽 탐색

def inorder(node): #중위
    if tree[node][0] != ".":
        inorder(tree[node][0]) #왼쪽 탐색
    print(node, end="")
    if tree[node][1] != ".":
        inorder(tree[node][1]) #오른쪽 탐색        
        
def postorder(node): #후위
    if tree[node][0] != ".":
        postorder(tree[node][0]) #왼쪽 탐색
    if tree[node][1] != ".":
        postorder(tree[node][1]) #오른쪽 탐색   
    print(node, end="")
    
preorder("A")
print()
inorder("A")
print()
postorder("A")